src/t8_forest/t8_forest_ghost.cxx

```diff
@@ -1929,4 +1880,348 @@ t8_forest_ghost_destroy (t8_forest_ghost_t *pghost)
+{
+  t8_global_productionf (" t8_forest_ghost_interface_faces::t8_ghost_step_1_clean_up \n");
+  if (memory_flag & CREATE_GFIRST_DESC_ARRAY) {
+    /* Free the offset memory, if created */
+    t8_shmem_array_destroy (&forest->element_offsets);
+  }
+  if (memory_flag & CREATE_TREE_ARRAY) {
+    /* Free the offset memory, if created */
+    t8_shmem_array_destroy (&forest->tree_offsets);
+  }
+  if (memory_flag & CREATE_GFIRST_DESC_ARRAY) {
+    /* Free the offset memory, if created */
+    t8_shmem_array_destroy (&forest->global_first_desc);
+  }
+}
+
+/**
+ * Derived class ghost_w_search
+*/
+
+t8_forest_ghost_w_search::t8_forest_ghost_w_search (t8_ghost_type_t ghost_type): t8_forest_ghost_interface (ghost_type)
+{
+  T8_ASSERT (ghost_type != T8_GHOST_NONE);
+  T8_ASSERT (ghost_type == T8_GHOST_FACES);  // currently no other typs are suportet
+  if (ghost_type == T8_GHOST_FACES) {
+    search_fn = t8_forest_ghost_search_boundary;
+  }
+  SC_CHECK_ABORT (ghost_type != T8_GHOST_USERDEFINED,
+                  "use t8_forest_ghost_w_search(t8_forest_search_fn search_function) for userdefined ghost");
+}
+
+void
+t8_forest_ghost_w_search::do_ghost (t8_forest_t forest)
+{
+
+  communicate_ownerships (forest);
+
+  if (t8_forest_get_local_num_elements (forest) > 0) {
+    if (t8_ghost_get_type () == T8_GHOST_NONE) {
+      t8_debugf ("WARNING: Trying to construct ghosts with ghost_type NONE. "
+                 "Ghost layer is not constructed.\n");
+      return;
+    }
+
+    /* Initialize the ghost structure */
+    t8_forest_ghost_init (&forest->ghosts, ghost_type);
+
+    search_for_ghost_elements (forest);
+
+    communicate_ghost_elements (forest);
+  }
+  clean_up (forest);
+}
+
+void
+t8_forest_ghost_w_search::search_for_ghost_elements (t8_forest_t forest)
+{
+  t8_forest_ghost_boundary_data_t data;
+  void *store_user_data = NULL;
+
+  /* Start with invalid entries in the user data.
+   * These are set in t8_forest_ghost_search_boundary each time a new tree is entered */
+  data.eclass = T8_ECLASS_COUNT;
+  data.gtreeid = -1;
+  data.ts = NULL;
+#ifdef T8_ENABLE_DEBUG
+  data.left_out = 0;
+#endif
+  sc_array_init (&data.face_owners, sizeof (int));
+  /* This is a dummy init, since we call sc_array_reset in ghost_search_boundary
+   * and we should not call sc_array_reset on a non-initialized array */
+  sc_array_init (&data.bounds_per_level, 1);
+  /* Store any user data that may reside on the forest */
+  store_user_data = t8_forest_get_user_data (forest);
+  /* Set the user data for the search routine */
+  t8_forest_set_user_data (forest, &data);
+  /* Loop over the trees of the forest */
+  t8_forest_search (forest, search_fn, NULL, NULL);
+
+  /* Reset the user data from before search */
+  t8_forest_set_user_data (forest, store_user_data);
+
+  /* Reset the data arrays */
+  sc_array_reset (&data.face_owners);
+  sc_array_reset (&data.bounds_per_level);
+#ifdef T8_ENABLE_DEBUG
+#endif
```

Comment:
?

Reply:
This is 8 years old by Johannes. but yes I thing this could removed. Should we ask Johannes if there is a real reason for this?

Reply:
No, just remove it